### ISSUE TITLE
date_histogram's offset doesn't like daylight savings time

### ISSUE THREAD

**Issue by nik9000** [elastic/elasticsearch#56305]
`date_histogram`'s [offset](https://www.elastic.co/guide/en/elasticsearch/reference/7.8/search-aggregations-bucket-datehistogram-aggregation.html#search-aggregations-bucket-datehistogram-offset) advertises itself as shifting around the start of the histogram interval. It says:

```
For example, when using an interval of day, each bucket runs from midnight to midnight. Setting the offset parameter to +6h changes each bucket to run from 6am to 6am
```

This is mostly true but not actually true. What it really does is shift each document's time back the offset, perform the rounding, and then shift it forward by the offset. If it weren't for daylight savings time this would do exactly as advertised. But, daylight savings time is hard. Check this out:
```
ES='curl -s -uelastic:password'
ES_DELETE="$ES -XDELETE"
ES_POST="$ES -XPOST -HContent-Type:application/json"
ES_PUT="$ES -XPUT -HContent-Type:application/json"

$ES_DELETE localhost:9200/test
$ES_PUT localhost:9200/test -d'{
  "settings": {
    "number_of_replicas": 0
  },
  "mappings": {
    "properties": {
      "timestamp": { "type": "date" }
    }
  }
}'

$ES_POST 'localhost:9200/test/_doc?pretty&refresh' -d'{"date": "2020-03-09T04:00:00Z"}'
$ES_POST 'localhost:9200/test/_doc?pretty&refresh' -d'{"date": "2020-03-09T05:00:00Z"}'
$ES_POST 'localhost:9200/test/_doc?pretty&refresh' -d'{"date": "2020-03-09T06:00:00Z"}'
$ES_POST 'localhost:9200/test/_doc?pretty&refresh' -d'{"date": "2020-03-09T07:00:00Z"}'
$ES_POST 'localhost:9200/test/_doc?pretty&refresh' -d'{"date": "2020-03-09T08:00:00Z"}'

$ES_POST 'localhost:9200/test/_search?size=0&pretty' -d'{
  "aggs": {
    "oh_no": {
      "date_histogram": {
        "field": "date",
        "calendar_interval": "1d",
        "time_zone": "America/New_York",
        "offset": "+1h"
      }
    }
  }
}'
```

That results in:
```
  "aggregations" : {
    "oh_no" : {
      "buckets" : [
        {
          "key_as_string" : "2020-03-08T00:00:00.000-05:00",
          "key" : 1583643600000,
          "doc_count" : 1
        },
        {
          "key_as_string" : "2020-03-08T01:00:00.000-05:00",
          "key" : 1583647200000,
          "doc_count" : 0
        },
        {
          "key_as_string" : "2020-03-09T01:00:00.000-04:00",
          "key" : 1583730000000,
          "doc_count" : 4
        }
      ]
    }
  }
```

`2020-03-08T00:00:00.000-05:00` doesn't start at 1am as advertised. Ooops. `2020-03-08T01:00:00.000-05:00` *does* start at 1am but it is empty. Oh no. Something is all wrong.....

The offset implementation that we have just isn't going to work properly with daylight savings time. We really should convert to and from local time before applying the offset. Or something like that.......

**Comment by salvatore-campagna** [comment]
According to tests I wrote in #86526 I cannot reproduce the issue. Will try to see if I can identify when it was fixed (assuming the test I wrote are correct).

**Comment by salvatore-campagna** [comment]
Fixed before `7.17.0`
